// ng-mat-select-infinite-scroll/__ivy_ngcc__/fesm2015/ng-mat-select-infinite-scroll.js
import { EventEmitter, Directive, NgZone, Input, Output, NgModule } from '@angular/core';
import { SELECT_ITEM_HEIGHT_EM, MatSelect, MatSelectModule } from '@angular/material/select';
import { takeUntil, debounceTime, tap } from 'rxjs/operators';
import { Subject, fromEvent } from 'rxjs';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/material/select';
class MatSelectInfiniteScrollDirective {
    /**
     * @param {?} matSelect
     * @param {?} ngZone
     */
    constructor(matSelect, ngZone) {
        this.matSelect = matSelect;
        this.ngZone = ngZone;
        this.threshold = '15%';
        this.debounceTime = 150;
        this.infiniteScroll = new EventEmitter();
        this.thrPx = 0;
        this.thrPc = 0;
        this.singleOptionHeight = SELECT_ITEM_HEIGHT_EM;
        this.destroyed$ = new Subject();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.evaluateThreshold();
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.matSelect.openedChange.pipe(takeUntil(this.destroyed$)).subscribe((/**
         * @param {?} opened
         * @return {?}
         */
        (opened) => {
            if (opened) {
                this.panel = this.matSelect.panel.nativeElement;
                this.singleOptionHeight = this.getSelectItemHeightPx();
                this.registerScrollListener();
            }
        }));
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.destroyed$.next(true);
        this.destroyed$.complete();
    }
    /**
     * @return {?}
     */
    evaluateThreshold() {
        if (this.threshold.lastIndexOf('%') > -1) {
            this.thrPx = 0;
            this.thrPc = (parseFloat(this.threshold) / 100);
        }
        else {
            this.thrPx = parseFloat(this.threshold);
            this.thrPc = 0;
        }
    }
    /**
     * @return {?}
     */
    registerScrollListener() {
        fromEvent(this.panel, 'scroll').pipe(takeUntil(this.destroyed$), debounceTime(this.debounceTime), tap((/**
         * @param {?} event
         * @return {?}
         */
        (event) => {
            this.handleScrollEvent(event);
        }))).subscribe();
    }
    /**
     * @param {?} event
     * @return {?}
     */
    handleScrollEvent(event) {
        this.ngZone.runOutsideAngular((/**
         * @return {?}
         */
        () => {
            if (this.complete) {
                return;
            }
            /** @type {?} */
            const countOfRenderedOptions = this.matSelect.options.length;
            /** @type {?} */
            const infiniteScrollDistance = this.singleOptionHeight * countOfRenderedOptions;
            /** @type {?} */
            const threshold = this.thrPc !== 0 ? (infiniteScrollDistance * this.thrPc) : this.thrPx;
            /** @type {?} */
            const scrolledDistance = this.panel.clientHeight + event.target.scrollTop;
            if ((scrolledDistance + threshold) >= infiniteScrollDistance) {
                this.ngZone.run((/**
                 * @return {?}
                 */
                () => this.infiniteScroll.emit()));
            }
        }));
    }
    /**
     * @return {?}
     */
    getSelectItemHeightPx() {
        return parseFloat(getComputedStyle(this.panel).fontSize) * SELECT_ITEM_HEIGHT_EM;
    }
}
MatSelectInfiniteScrollDirective.ɵfac = function MatSelectInfiniteScrollDirective_Factory(t) { return new (t || MatSelectInfiniteScrollDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.MatSelect), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.NgZone)); };
MatSelectInfiniteScrollDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: MatSelectInfiniteScrollDirective, selectors: [["", "msInfiniteScroll", ""]], inputs: { threshold: "threshold", debounceTime: "debounceTime", complete: "complete" }, outputs: { infiniteScroll: "infiniteScroll" } });
/** @nocollapse */
MatSelectInfiniteScrollDirective.ctorParameters = () => [
    { type: MatSelect },
    { type: NgZone }
];
MatSelectInfiniteScrollDirective.propDecorators = {
    threshold: [{ type: Input }],
    debounceTime: [{ type: Input }],
    complete: [{ type: Input }],
    infiniteScroll: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(MatSelectInfiniteScrollDirective, [{
        type: Directive,
        args: [{
                selector: '[msInfiniteScroll]'
            }]
    }], function () { return [{ type: ɵngcc1.MatSelect }, { type: ɵngcc0.NgZone }]; }, { threshold: [{
            type: Input
        }], debounceTime: [{
            type: Input
        }], infiniteScroll: [{
            type: Output
        }], complete: [{
            type: Input
        }] }); })();
if (false) {
    /** @type {?} */
    MatSelectInfiniteScrollDirective.prototype.threshold;
    /** @type {?} */
    MatSelectInfiniteScrollDirective.prototype.debounceTime;
    /** @type {?} */
    MatSelectInfiniteScrollDirective.prototype.complete;
    /** @type {?} */
    MatSelectInfiniteScrollDirective.prototype.infiniteScroll;
    /**
     * @type {?}
     * @private
     */
    MatSelectInfiniteScrollDirective.prototype.panel;
    /**
     * @type {?}
     * @private
     */
    MatSelectInfiniteScrollDirective.prototype.thrPx;
    /**
     * @type {?}
     * @private
     */
    MatSelectInfiniteScrollDirective.prototype.thrPc;
    /**
     * @type {?}
     * @private
     */
    MatSelectInfiniteScrollDirective.prototype.singleOptionHeight;
    /**
     * @type {?}
     * @private
     */
    MatSelectInfiniteScrollDirective.prototype.destroyed$;
    /**
     * @type {?}
     * @private
     */
    MatSelectInfiniteScrollDirective.prototype.matSelect;
    /**
     * @type {?}
     * @private
     */
    MatSelectInfiniteScrollDirective.prototype.ngZone;
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class MatSelectInfiniteScrollModule {
}
MatSelectInfiniteScrollModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: MatSelectInfiniteScrollModule });
MatSelectInfiniteScrollModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function MatSelectInfiniteScrollModule_Factory(t) { return new (t || MatSelectInfiniteScrollModule)(); }, imports: [[
            MatSelectModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(MatSelectInfiniteScrollModule, { declarations: function () { return [MatSelectInfiniteScrollDirective]; }, imports: function () { return [MatSelectModule]; }, exports: function () { return [MatSelectInfiniteScrollDirective]; } }); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(MatSelectInfiniteScrollModule, [{
        type: NgModule,
        args: [{
                declarations: [MatSelectInfiniteScrollDirective],
                imports: [
                    MatSelectModule
                ],
                exports: [MatSelectInfiniteScrollDirective]
            }]
    }], null, null); })();

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { MatSelectInfiniteScrollDirective, MatSelectInfiniteScrollModule };
